Introducing SAUL, the [S]ensor [A]ctuator [U]ber [L]ayer

haukepetersen opened the pull request:
~~Let me introduce to you RASI - the all new generic [R]IOT [A]ctuator [S]ensor [I]nterface.~~
**EDIT** Let me introduce to you SAUL - the all new generic [S]ensor [A]ctuator [U]ber [L]ayer.
 I had a nice idea for making all sensors and actuator in a system available via a unified API, and here are the results.

The concepts are very similar to (grnc) netdev and netif - every actuator and sensor is accessible through the same API (read and write for now) with one single data structure used of parameter passing -> `phydat_t`. All devices register at a central location and can then be listed, queried, searched and accessed. By specifying unique classes of devices, one could search for: "give me all sensors, that read pressure values" or similar.

This data structure `phydat_t` is made for carrying up to 3 dimensions of physical values, e.g. values read by sensor or set values for actuators. It takes only ~8 byte and contains additional to the data the physical unit and a scaling factor.

Also I included the sensors for auto-initialization, exactly in the same way we do it for network interfaces: In the board or in the application you just define a set of parameters for a device, and auto-init takes care of the rest.

Testing this PR is best done on the `iot-lab_M3`, as I included all 4 sensors and 3 LED (actuators) into this PR. A little comparison:

Code size of the default example before:
```
   text	   data	    bss	    dec	    hex
  28592	    168	   7304	  36064	   8ce0
```
and after:
```
   text	   data	    bss	    dec	    hex
  27648	    168	   7408	  35224	   8998
```
We save ~900 byte of flash and lose ~100byte of RAM. The saved flash is mainly due to one unified shell command (instead of one for each sensor before). The lost RAM is spend on the RASI registry, in this setup there are 8 devices in it (5 sensor entries, 3 LEDs) and one entry is 12 byte...

EDIT
~~The amount of touched files looks worse than it is - this PR is based on #3535 and #3420.~~
Removed all dependencies

authmillenon commented:
You might want to overthink [that name](https://en.wikipedia.org/wiki/Golden_Raspberry_Awards) ;-)

haukepetersen commented:
possibly... Any better ideas?

authmillenon commented:
(U)ASAPI [(**u**nified) **a**ctuator **s**ensor **API**]?

authmillenon commented on sys/include/phydat.h:
```diff
@@ -0,0 +1,165 @@
+    UNIT_TIME,      /**< the three dimensions contain sec, min, and hours */
+    UNIT_DATE,      /**< the 3 dimensions contain days, months and years */
+    /* extend this list as needed */
+};
```
Why no typedef for that? `phydat_unit_t` e.g.

authmillenon replied:
As an ex-chemist I see a worrying lack of `UNIT_MOL` ;-P

authmillenon replied:
(or as a more practical approach `UNIT_PERCENT`, `UNIT_PERMILLE`, and `UNIT_PPM`)

haukepetersen replied:
As stated in the comment, extend the list when you need anything...

authmillenon commented on sys/include/phydat.h:
```diff
@@ -0,0 +1,165 @@
+ */
+typedef struct {
+    int16_t val[PHYDAT_DIM];    /**< the 3 generic dimensions of data */
+    uint8_t unit;               /**< the (physical) unit of the data */
```
Ah… I think I got it :D (so that you don't have to use it here as a word-sized type ;-))

haukepetersen replied:
yepp

authmillenon commented on sys/include/phydat.h:
```diff
@@ -0,0 +1,165 @@
+ * @param[in] data      data container to dump
+ * @param[in] dim       number of dimension of @p data to dump
+ */
+void phydat_dump(phydat_t *data, int dim);
```
For usage with CBOR, JSON etc. an intermediate function (e.g. `phydat_str()`) that converts the same output just to a string might be of use.

haukepetersen replied:
I already had the prototype of such a function written down, but dropped it for later...

authmillenon commented on sys/phydat/phydat_str.c:
```diff
@@ -0,0 +1,79 @@
+        return;
+    }
+    printf("Data:");
+    for (int i = 0; i < dim; i++) {
```
From the doc I would have thought that `dim` is the index, not the maximum for the indices.

haukepetersen replied:
the maximum is `PHYDAT_DIM`. `dim` as paramter is used so you can print only a subset, e.g. 1-3, of the values - as not always all of them are populated

authmillenon commented on sys/phydat/phydat_str.c:
```diff
@@ -0,0 +1,79 @@
+        case 6:     return "M";
+        case 9:     return "G";
+        case 12:    return "T";
+        default:    return "*10^X";     /* TODO: how to output this here */
```
Make the string an `out` parameter `str` and it's maximum length a precondition:

```C
sprintf(str, "*e%" PRIi8 " ", scale);
```

haukepetersen replied:
But that makes the function much less versatile to use... And after all, it is just a print function for debugging/shell usage...

authmillenon replied:
Is it though? I can imagine using this to be send directly over CoAP as output data.

Kijewski commented on sys/phydat/phydat_str.c:
```diff
@@ -0,0 +1,79 @@
+    switch (unit) {
+        case UNIT_DEG_C:    return "°C";
+        case UNIT_DEG_F:    return "°F";
+        case UNIT_DEG_K:    return "°K";
```
"Unlike the degree Fahrenheit and degree Celsius, the kelvin is not referred to or typeset as a degree"

haukepetersen replied:
You are right, thats what happens once you are in a flow :-) Will fix.

haukepetersen commented:
Thought about the name,  how about `arcade`? just not sure about the meaning, maybe something like Awesome RIOT Control/Acquisition Device Extension? Ideas?

kaspar030 commented:
Nice work!

I think we should try to get away with a static (compile-time)registry., though. Similar to how we did it for shell commands.

OlegHahm commented:
> I think we should try to get away with a static (compile-time)registry., though. Similar to how we did it for shell commands.

:+1:  - and once again I'm not in favor of introducing more function pointers. I find them super annoying for debugging.

haukepetersen commented:
Actually I am against the static approach, as this does not allow for hot-plugging - and that is a feature that I see as quite interesting and more-then-nice-to-have for the future!

haukepetersen commented:
Renamed the interface, now i's all good man :-)

Also I separated the actual interface (`drivers/saul`) from the device registry (`sys/saul_reg`). For the registry, I switched from a pre-allocated array to a linked list. Although the lookups are slightly less efficient, the memory handling is way easier. Also the lookup mostly only has to be done once and afterwards a module can interact with a device directly (exception are shell commands, but for these performance is not the top-most concern...).

@kaspar030, care to take a look?

haukepetersen commented:
rebased

haukepetersen commented:
Reworked this PR:
- removed all dependencies to other PRs
- repaired some minor things
- changed the `phydat_scale_to_str()` to write in a given buffer
- separated the `name` parameter for the `saul_reg` from the sensor's `xx_param_t` structs

So IMHO this should be read for review/merge - let's go :-)

kaspar030 commented on sys/phydat/phydat_str.c:
```diff
@@ -0,0 +1,80 @@
+        case 12:    str[0] = 'T'; break;
+        case 15:    str[0] = 'P'; break;
+        default:
+            snprintf(str, PHYDAT_SCALE_STR_MAXLEN, "E%i", (int)scale);
```
could you use fmt_* for this?

haukepetersen replied:
I don't know, probably if you say so :-) Will look into this

haukepetersen replied:
But then again: why? Just tried it: 
```c
        default: {
            str[0] = 'E';
            size_t pos = fmt_s32_dec(&str[1], (int32_t)scale);
            str[pos + 1] = '\0';
            return;
        }
```
This (i) pulls in the `fmt` module and (ii) increases the code size by 24 byte - all for a function, which major use is for debugging/shell...

kaspar030 replied:
Check https://github.com/haukepetersen/RIOT/pull/21. It saves 8 bytes on samr21, even counting the inclusion of fmt. Your function gets a little smaller. The whole idea is to not need *printf.

(If you agree, don't merge my PR, but cherry pick the phydat commit, the other is PRed in #4306).

haukepetersen replied:
this does indeed make sense then - agree :-)

kaspar030 commented on sys/saul_reg/saul_reg.c:
```diff
@@ -0,0 +1,133 @@
+
+    while (tmp) {
+        if (strcmp(tmp->name, name) == 0) {
+            return tmp;
```
break;

haukepetersen replied:
what is the advantage of `break` here?

kaspar030 replied:
combined with "return tmp" downwards, IMHO it leads to smaller code, as e.g., "NULL" doesn't have to be stored as literal, and "tmp" is kept in register anyways.

But sorry, that was untested premature optimization. ;)

haukepetersen replied:
One simple solution: let's test it!

haukepetersen replied:
As (I actually) expected:
current state:   
```
text	   data	    bss	    dec	    hex	filename
34180	    164	  11772	  46116	   b424
```
you proposal:
```
   text	   data	    bss	    dec	    hex	filename
  34180	    164	  11772	  46116	   b424
```
so I guess the compiler is smart enough in this case... Are you ok with leaving this as is then?

kaspar030 replied:
yup! :=)

kaspar030 commented on sys/saul_reg/saul_reg.c:
```diff
@@ -0,0 +1,133 @@
+        }
+        tmp = tmp->next;
+    }
+    return NULL;
```
return tmp;

kaspar030 commented:
Yay, first code using fmt! :)

kaspar030 commented:
(This probably doesn't compile on avr and msp430 without #4312).

haukepetersen commented:
damn, then I guess we have to wait for #4312. But lets try to pull this in as soon as possible, I want SAUL merged!

kaspar030 commented:
please rebase!

kaspar030 commented:
#4312 is merged.

haukepetersen commented:
rebased

haukepetersen commented:
@kaspar030, should I squash?

kaspar030 commented on boards/iotlab-m3/include/gpio_params.h:
```diff
@@ -0,0 +1,56 @@
+#endif
+
+/**
+ * @brief    LPS331AP configuration
```
gpio pin configuration?

kaspar030 commented on boards/iotlab-m3/include/gpio_params.h:
```diff
@@ -0,0 +1,56 @@
+}
+#endif
+
+#endif /* LPS331AP_PARAMS_H */
```
GPIO_*

kaspar030 commented:
See https://github.com/haukepetersen/RIOT/pull/22.

Works fine for setting the led, just this comes to mind:

When using "saul write" to set my LED, "saul write 0 0" turns it on, "0 1" turns it off. While that is correct in respect to the LED's wiring, it seems unexpected. Maybe a flag to invert the gpio values would help. But I guess that can come later...

kaspar030 commented:
Can't test the other drivers for lack of hardware, but I'd say, please squash and merge my commit!

kaspar030 commented:
BTW, this is another candidate for heavy simplification using #4290. For most use-cases, the run-time registry can be replaced by static compile-time configuration.

kaspar030 commented:
BTW, I already like SAUL a lot. IMHO the best new feature in a while.

haukepetersen commented:
haukepetersen#22 is merged, will fix and squash after our meeting now.

kaspar030 commented on boards/fox/include/l3g4200d_params.h:
```diff
@@ -0,0 +1,58 @@
+ * @author    Hauke Petersen <hauke.petersen@fu-berlin.de>
+ */
+
+#ifndef LPS331AP_PARAMS_H
```
define correct?

kaspar030 commented on boards/fox/include/l3g4200d_params.h:
```diff
@@ -0,0 +1,58 @@
+}
+#endif
+
+#endif /* LPS331AP_PARAMS_H */
```
here

kaspar030 commented on boards/iotlab-m3/include/l3g4200d_params.h:
```diff
@@ -0,0 +1,59 @@
+ * @author    Hauke Petersen <hauke.petersen@fu-berlin.de>
+ */
+
+#ifndef LPS331AP_PARAMS_H
```
...

kaspar030 commented on boards/iotlab-m3/include/l3g4200d_params.h:
```diff
@@ -0,0 +1,59 @@
+}
+#endif
+
+#endif /* LPS331AP_PARAMS_H */
```
...

haukepetersen commented:
squashed

OlegHahm commented on drivers/include/saul.h:
```diff
@@ -0,0 +1,157 @@
+ * @ingroup     drivers
+ * @brief       Generic sensor/actuator abstraction layer for RIOT
+ *
+ * SAUL is a generic actuator/sensor interface in RIOT. It's purpose is to
```
`s/It's/Its/`

OlegHahm commented on drivers/include/saul.h:
```diff
@@ -0,0 +1,157 @@
+ * enable unified interaction with a wide range of sensors and actuators through
+ * a set of defined access functions and a common data structure.
+ *
+ * Each device implementing this interface has to expose a set of predefined
```
`s/device/device driver/`

OlegHahm commented on drivers/include/saul.h:
```diff
@@ -0,0 +1,157 @@
+    SAUL_SENSE_COLOR    = 0x88,     /**< sensor: (light) color */
+    SAUL_SENSE_PRESS    = 0x89,     /**< sensor: pressure */
+    SAUL_SENSE_ANALOG   = 0x8a,     /**< sensor: raw analog value */
+    SAUL_CLASS_ANY      = 0xff,     /**< any device - wildcard */
```
Remove the last comma to be pedantic- and C++ compliant.

see https://github.com/RIOT-OS/RIOT/pull/4358#discussion_r46080465

OlegHahm commented on sys/include/phydat.h:
```diff
@@ -0,0 +1,175 @@
+    UNIT_PPM,       /**< part per million */
+    /* aggregate values */
+    UNIT_TIME,      /**< the three dimensions contain sec, min, and hours */
+    UNIT_DATE,      /**< the 3 dimensions contain days, months and years */
```
see above

OlegHahm commented on sys/include/phydat.h:
```diff
@@ -0,0 +1,175 @@
+    /* temperature */
+    UNIT_DEG_C,     /**< degree Celsius */
+    UNIT_DEG_F,     /**< degree Fahrenheit */
+    UNIT_DEG_K,     /**< degree Kelvin */
```
nitpicky, but there is no such thing as "degree Kelvin".

OlegHahm commented on sys/include/saul_reg.h:
```diff
@@ -0,0 +1,144 @@
+ */
+typedef struct {
+    const char *name;           /**< string identifier for a device */
+} saul_reg_info_t;
```
what is the difference to `saul_reg_t::name`?

haukepetersen replied:
no same name (both point to the same text most likely), `saul_reg_t` is located in RAM for holding a `saul reg` entry, `saul_reg_info_t` is used for defining additional information about a device, e.g. used by auto-init

haukepetersen replied:
At some point we probably may decide to either get rid of this type and use a string directly or to let `saul_reg_t` point to this, dependent if there will be more information (fields) that is added to `sauL_reg_info_t`

OlegHahm commented:
Reviewed the latest commits fixing my minor remarks. Please squash.

haukepetersen commented:
squashed

haukepetersen commented:
Travis is happy - @kaspar030: do you ack?

kaspar030 commented:
Yep. go!

OlegHahm commented:
Congrats!

sgso commented:
Is this suitable to replace the awful LED-Macros some tests depend on? Instead of relying on LED_RED and other stuff to be defined (even if there aren't any red ones ;) use the 1st, 2nd, etc. actuator of type led?

OlegHahm commented:
For LEDs I think there's no way around macros, because there are often inevitable for debugging where any overhead is not tolerable. I wouldn't object renaming to something like `LED_DEBUG1` or the like, though.

daniel-k commented on sys/include/phydat.h:
```diff
@@ -0,0 +1,175 @@
+/**
+ * @brief   Generic data structure for expressing physical values
+ *
+ * Physical data is express in a 3-dimensional touple of values. In addition to
```
`s/express/expressed/`

Sorry for being that late, I waited with reading the docs until now that SAUL is merged.

daniel-k commented on sys/include/phydat.h:
```diff
@@ -0,0 +1,175 @@
+ * is expressed as power of 10 (10^factor).
+ *
+ * The combination of signed 16-bit numbers with and the scale factor gives us a
+ * very high dynamic range (from -32*10^-131 to 32*10^130). I a wider sense we
```
`s/I a wider/In a wider/`

haukepetersen commented:
good catches -> fixed in #4369
